File: system-test/java/src/main/java/com/mycompany/myshop/testkit/dsl/core/shared/UseCaseResult.java
package com.mycompany.myshop.testkit.dsl.core.shared;

import com.mycompany.myshop.testkit.driver.port.dtos.error.SystemError;
import com.mycompany.myshop.testkit.common.Result;

import java.util.function.BiFunction;

import static com.mycompany.myshop.testkit.common.ResultAssert.assertThatResult;

public class UseCaseResult<R, V> {
    private final Result<R, SystemError> result;
    private final UseCaseContext context;
    private final BiFunction<R, UseCaseContext, V> successVerificationFactory;

    public UseCaseResult(
            Result<R, SystemError> result,
            UseCaseContext context,
            BiFunction<R, UseCaseContext, V> successVerificationFactory) {
        this.result = result;
        this.context = context;
        this.successVerificationFactory = successVerificationFactory;
    }

    public V shouldSucceed() {
        assertThatResult(result).isSuccess();
        return successVerificationFactory.apply(result.getValue(), context);
    }

    public ErrorVerification shouldFail() {
        assertThatResult(result).isFailure();
        return new ErrorVerification(result.getError(), context);
    }
}

File: system-test/typescript/src/testkit/dsl/core/shared/use-case-result.ts
import type { Result } from '../../../common/result.js';
import type { SystemError } from '../../../driver/port/dtos/errors/SystemError.js';
import { ErrorVerification } from './error-verification.js';
import type { UseCaseContext } from './use-case-context.js';

export type SuccessVerificationFactory<TResponse, TVerification> = (
  response: TResponse,
  context: UseCaseContext,
) => TVerification;

export class UseCaseResult<TResponse, TSuccessVerification> {
  constructor(
    private readonly result: Result<TResponse, SystemError>,
    private readonly context: UseCaseContext,
    private readonly successVerificationFactory: SuccessVerificationFactory<TResponse, TSuccessVerification>,
  ) {}

  shouldSucceed(): TSuccessVerification {
    if (!this.result.success) {
      throw new Error(`Expected success but was failure: ${JSON.stringify(this.result.error)}`);
    }
    return this.successVerificationFactory(this.result.value, this.context);
  }

  shouldFail(): ErrorVerification {
    if (this.result.success) {
      throw new Error(`Expected failure but was success: ${JSON.stringify(this.result.value)}`);
    }
    return new ErrorVerification(this.result.error, this.context);
  }
}
